plugins/bil/bil_main.ml

```diff
@@ -34,44 +33,46 @@ let eliminate_dead_code bil =
       | st -> loop (st :: acc) bil in
   loop [] bil
 
-let apply bil =
-  List.fold ~init:bil ~f:(fun bil f -> f bil)
-
-let norml level =
-  if level = 0 then ident
-  else
-    Stmt.normalize ~keep_ites:false ~normalize_exp:false
+let () =
+  Bil.register_pass "normalization" @@
+  Stmt.normalize ~keep_ites:false ~normalize_exp:false
 
-let simpl level =
-  if level = 0 then ident
-  else
-    Bil.fixpoint
-      (fun bil -> Bil.fold_consts bil |> Bil_const_propagation.run |> eliminate_dead_code)
+let () =
+  Bil.register_pass "optimization" @@
+  Bil.fixpoint
+    (fun bil -> Bil.fold_consts bil |> Bil_const_propagation.run |> eliminate_dead_code)
 
-let run norml_level simpl_level =
-  provide_bil_transformation "internal" @@
-  fun bil -> apply bil [
-    norml norml_level;
-    simpl simpl_level;
-  ]
+let run passes =
+  let passes = ["normalization"; "optimization"] @ passes in
+  let is_selected p =
+    List.mem passes ~equal:String.equal (Bil.Pass.to_string p) in
+  let all = Bil.passes () in
+  List.filter all ~f:is_selected |>
+  Bil.select_passes
 
 let () =
   let () = Config.manpage [
       `S "DESCRIPTION";
-      `P "Applies analysises to a instruction bil code" ;
-      `Pre "
-The following bil analysises are in default pipeline:
-Constant Folding and Simplification
-Bil Normalization
-Constant Propagation
-";
-      `S "SEE ALSO";
-      `P "$(b,bap.mli)"
+      `P "Establishes a pipeline of BIL analysis";
```

Comment:
Use the following manpage:

Controls the BIL transformation pipeline. 

The BIL transformation pipeline is used to transform the BIL code of each machine instruction and is applied to the output of a lifter. Each piece of transformation is called $(i,a pass). This plugin enables full control over the pass selection as well as provides some predefined passes. See the $(b,--bil-list-passes) for a comprehensive list of
available transformations. The pipeline is formed as a concatenation of all normalization passes that correspond to the selected level of normalization (setting it to zero will disable normalizations), all optimization passes applicable to the selected level of optimization (again the zero level corresponds to an empty list of optimizations), and all passes selected via the `$(b,--bil-passes)` command line argument. All passes are applied until a fixed point is reached (i.e., until the outcome of the pipeline is equal to the income) or until a loop is detected. Thus to prevent infinite looping each registered pass should be a pure function of its inputs (same input shall result in the same output). 

A new pass could registered via the $(b,Bil.register_pass) function. It won't be automatically selected, so it should be specified explicitly via the $(b,--bil-passes) command line argument.